make_module: First version

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23288 make_module: First version**

Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
make_module: First version

zafartahirov commented:
Example:

```python
mod = make_module(torch.add)
mod_quant = mod.from_float()

a = torch.tensor(1, dtype=torch.float)
qa = torch.quantize_linear(a, 1.0, 127, torch.quint8)
b = torch.tensor(2, dtype=torch.float)
qb = torch.quantize_linear(b, 1.0, 127, torch.quint8)

print("Types\t\t:", type(mod), type(mod_quant))
print("__repr__\t:", mod, mod_quant)
print("forward call\t:", mod(a, b))
print("forward call\t:", mod_quant(qa, qb, 1.0, 0))
```

Prints:

```
Types		: <class '__main__._WrapModule'> <class '__main__._WrapModule'>
__repr__	: Add() Add_quantized()
forward call	: tensor(3.)
forward call	: tensor(3., size=(), dtype=torch.quint8, scale=1.0, zero_point=0)
```

raghuramank100 commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,61 @@
+    def forward(self, *args, **kwargs):
+        return self.func(*args, **kwargs)
+
+    def from_float(self, *args, **kwargs):
```
Great start!. from_float is implemented as a method in quantized modules to construct it from a float module. It is required for all quantized modules so that torch.quantization.convert can have a uniform API for module conversion. . In our implementations so far, from_float is a classmethod for modules in nn.quantized. (Please see: https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py for an example)

Please remove from_float() here and have it in the quantized module.
I would suggest having the following:
```PYTHON
# nn.quantized.Add
class Add(nn.Module):
      def __init__(self, out_scale=1.0, out_zero_point=0):
        super(Add, self).__init__()
        self.register_buffer('scale', torch.tensor(scale, dtype=torch.double))
        self.register_buffer('zero_point', torch.tensor(zero_point, dtype=torch.long))
     
     def forward(self, X,Y):
        return torch.ops.quantized.add(X, Y, scale =  self.scale,
                                     zero_point = self.zero_point)
        
    @staticmethod
    def from_float(mod):
        assert hasattr(mod, 'observer')
        scale, zp = mod.observer.calculate_qparams()
        return Add(scale, zp)
```

zafartahirov replied:
After discussing it with @jerryzh168, I think the current approach is better. The reason is that during the convert stage, we will have to (1) create additional classes (2) create mapping from `WrapModule` to those classes, (3) Search after the mapping which op ini the wrap module corresponds to which class in `nn.quantized`.

With the current approach, both quantized and non-quantized ops can be wrapped in the same class. At the same time, we need to maintain only one dict which maps an op to its quantied counterpart.

raghuramank100 replied:
I agree, let us do code-gen for both quantized and original modules for functions

raghuramank100 commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,61 @@
+    name: Name of the module class. If None, the class name will be capitalized
+          op name.
+"""
+class _WrapModule(torch.nn.Module):
```
With this approach, type(mod) will be <class '__main__._WrapModule>. An approach (inspired from Dima's earlier suggestion is to do codegen:

```python
import torch
import torch.nn as nn

def wrapper_forward(self, *args, **kwargs):
        # Expand tuple
        return self.operation(*args, **kwargs)

def create_module(operation):
    C= type(
        'torch_ops_'+operation.__name__,
        (nn.Module,),
        {
            "forward": wrapper_forward,
            "operation": operation,
        }
    )
    # Specify the type in globals()
    globals()[C.__name__] = C
    return C   

C1 = create_module(torch.add)()
print(C1.operation)
print(C1(2,3))
print(type(C1))
if type(C1) == torch_ops_add:
    print('Yes')
```  
Prints:
<built-in method add of type object at 0x7f4336df1a60>
tensor(5)
<class '__main__.torch_ops_add'>
Yes

raghuramank100 replied:
cc @dzhulgakov

zafartahirov replied:
There is another approach that we extensively discussed with @jerryzh168 yesterday. It achieves the same thing, and does not involce the use of globals:

```python
"""Factory method to create a module wrapper.

Args:
    func: The function to wrap the module around
    name: Name of the module class. If None, the class name will be capitalized op name.

Returns:
    Module class
"""
def make_module(op, quant_op=None, name=None):
    name = op.__name__.capitalize() if name is None else name
    
    if quant_op is None:
        quant_op = func_from_float.get(op, None)
    
    def init(self, func):
        super(type(self), self).__init__()
        self.func = func
        
    def forward(self, *args, **kwargs):
        return self.func(*args, **kwargs)
    
    def from_float(self, *args, **kwargs):
        return make_module(quant_op, name=name + "_quantized")

    module = type(name, 
                 (torch.nn.Module,),
                 {"__init__": init,
                    "forward": forward,
                    "from_float": from_float})
    return module(op)
```

This prints: 

```python
mod = make_module(torch.add)
mod_quant = mod.from_float()

a = torch.tensor(1, dtype=torch.float)
qa = torch.quantize_linear(a, 1.0, 127, torch.quint8)
b = torch.tensor(2, dtype=torch.float)
qb = torch.quantize_linear(b, 1.0, 127, torch.quint8)

print("Types\t\t:", type(mod), type(mod_quant))
print("__repr__\t:", mod, mod_quant)
print("forward call\t:", mod(a, b))
print("forward call\t:", mod_quant(qa, qb, 1.0, 0))
```

```
Types		: <class '__main__.Add'> <class '__main__.Add_quantized'>
__repr__	: Add() Add_quantized()
forward call	: tensor(3.)
forward call	: tensor(3., size=(), dtype=torch.quint8, scale=1.0, zero_point=0)
```

raghuramank100 reviewed (changes requested):
Please see suggestions

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

raghuramank100 commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+        return self.operation(*args, **kwargs)
+
+    def _from_float(self, mod, *args, **kwargs):
+        return make_module(mod.op, name=name, quantized=True)
```
Need to add logic to compute and assign scale and zero_point.

raghuramank100 replied:
Suggest creating two functions, one for float and one for quantized: 
```python
def make_module(op):
   name = op.__name__.capitalize()
   name = "_ops_" + name

    def _init(self, operation):
        super(type(self), self).__init__()
        self.operation = operation
        setattr(torch, name, type(self))  # Add the type to the torch namespace

  def _forward(self, *args, **kwargs):
        return self.operation(*args, **kwargs)

    _methods = {
        "__doc__": r"""Wrapper module around function {}.""".format(type(op)),
        "__init__": _init,
        "forward": _forward
    }
   module = type(name, (torch.nn.Module,), _methods)
   return module(op)
```
and 
```
def make_quantized_module(op):
   name = op.__name__.capitalize()
   name = "_ops_" + name+"_quantized"

    def _init(self, operation):
        super(type(self), self).__init__()
        self.operation = operation
        self.scale = 1.0
        self.zero_point = 0
        setattr(torch, name, type(self))  # Add the type to the torch namespace

  def _forward(self, *args, **kwargs):
        return self.operation(*args, **kwargs, scale = self.scale, zp = self.zero_point)

    _methods = {
        "__doc__": r"""Wrapper module around function {}.""".format(type(op)),
        "__init__": _init,
        "forward": _forward
    }
   module = type(name, (torch.nn.Module,), _methods)
   return module(op)
```

zafartahirov replied:
Should I just generate the default values, or should I extract it from somewhere?

zafartahirov replied:
Rewrote the whole thing -- now a set of files are generated based on the implemented ops

raghuramank100 commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+        "forward": _forward
+    }
+    if quantized:
+        _methods["from_float"] = _from_float
```
Need to add the attributes scale and zero_point to the module only if quantized.  Please add:

raghuramank100 reviewed (changes requested):
First draft, will add more comments later.

dzhulgakov commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+
+    if quantized:
+        name = name = name + "_quantized"
+        op = _operation_from_float.get(op, op)
```
you need to produce a nicer error message if the op is not found

zafartahirov replied:
So is this? Because we won't be able to generate the files if there are no ops.

dzhulgakov commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+    def _init(self, operation):
+        super(type(self), self).__init__()
+        self.operation = operation
+        setattr(torch, name, type(self))  # Add the type to the torch namespace
```
why do we need to do this?

raghuramank100 replied:
This is done to ensure that the type() returns a meaningful name for the created modules.

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -276,5 +279,8 @@ def swap_module(mod, mapping):
     if hasattr(mod, 'observer'):
         if type(mod) in mapping:
             new_mod = mapping[type(mod)].from_float(mod)
+        elif hasattr(mod, 'operation'):
```
have a base class for these wrappers instead?

dzhulgakov commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+    if quantized:
+        _methods["from_float"] = _from_float
+
+    module = type(name, (torch.nn.Module,), _methods)
```
does this magic work with TorchScript?

zafartahirov replied:
Now it does :)

dzhulgakov commented on torch/quantization/make_module.py:
```diff
@@ -0,0 +1,50 @@
+import torch
+
+_operation_from_float = {
```
if those are the only ops you support - why not just create classes explicitly at initialization instead of lazy creating them on demand? (you can still use for loop iteration). That way at least you'd get working pickle behavior and more hope to make things work with TorchScript.

Also, how many layers to you expect to have here? Many of the torch.nn.functional routines already have hard-coded Module equivalents, e.g. ReLU, Pad, etc.

raghuramank100 replied:
It is crucial to get things working with torchscript, so please move this code to init. The list  of ops that we would need is add, cat, mul.   We can start with add and cat for now.

zafartahirov replied:
I think this comment is now obsolete?

dzhulgakov reviewed (commented):
My biggest comment is that we need to figure out how to make it work with TorchScript (as you want the resulting model to be scriptable and runnable without python). Given that codegen + script to codegen might be an more robust approach (if you have a closed set of modules to cover which appears to be the case)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

zafartahirov commented on torch/quantization/make_modules.py:
```diff
@@ -0,0 +1,85 @@
+import os
```
This file is just a utility to generate the modules easier -- in case we need to generate a batch of modules at once

dzhulgakov replied:
Put this file into something more descriptive (such that it's not a normal module that can be imported) or better yet - move tools/build where other codegen recides

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

raghuramank100 commented on torch/quantization/make_modules.py:
```diff
@@ -0,0 +1,100 @@
+import os
+
+_disclaimer = r"""
+# This file is generated using `torch.nn.quantization.make_module`
```
This comment should not be present in this file

zafartahirov replied:
Actually, I think it should be. We need to clearly mark the files that can be overwritten as generated. Otherwise, a de developer can just modify them, and next time we run the generation, all changes will be lost

raghuramank100 commented on torch/quantization/make_modules.py:
```diff
@@ -0,0 +1,100 @@
+
+_module_init_extras = r'''
+        self.scale = kwargs.get('scale', 1.0)
+        self.zero_point = swargs.get('zero_point', 0)
```
Typo: kwargs

dzhulgakov commented on torch/quantization/_generated/QAddGen.py:
```diff
@@ -0,0 +1,33 @@
+import torch
+
+r"""AddGen wraps the torch.add function."""
+class AddGen(torch.nn.Module):
```
what's the point of calling it AddGen and then renaming it? Let's just give it the right name from the beginning

dzhulgakov commented on torch/quantization/_generated/QAddGen.py:
```diff
@@ -0,0 +1,33 @@
+
+r"""AddGen wraps the torch.add function."""
+class AddGen(torch.nn.Module):
+    def __init__(self, **kwargs):
```
why kwags?

zafartahirov replied:
To be able to reuse it for different types of ops that could be generated -- it's part of the template.

dzhulgakov commented on torch/quantization/_generated/QAddGen.py:
```diff
@@ -0,0 +1,33 @@
+    def __init__(self, **kwargs):
+        super(QAddGen, self).__init__()
+
+        self.scale = kwargs.get('scale', 1.0)
```
why kwags - can't they be regular args?

zafartahirov replied:
Because during the generation I am reusing the `init` part, and in non-quantized modules we don't have the scale and zero_point, they might or might not exist, depending on what is being generated.

dzhulgakov commented on torch/quantization/_generated/QAddGen.py:
```diff
@@ -0,0 +1,33 @@
+
+    @classmethod
+    def from_float(cls, mod):
+        assert hasattr(mod, 'observer')
```
don't you need to enforce the base to be Add?

dzhulgakov commented on torch/quantization/make_modules.py:
```diff
@@ -0,0 +1,100 @@
+    name = op_string.split('.')[-1].capitalize()
+    name = name + "Gen"
+
+    filename = 'Q' + name + '.py'
```
maybe just dump them into a single file? would be easier to import. And probably I wouldn't put it under _generated so that the names are nicer

zafartahirov replied:
The `_generated` namespace will not be accessible to the user. All the ops will be exposed only through the `quantization` namespace. Or should I put them into their own?

raghuramank100 replied:
Why not have the generated modules show up in two locations:
torch/nn/modules/generated (for float) and torch/nn/quantized/modules/generated (for quantized)?

dzhulgakov commented on torch/quantization/_generated/QAddGen.py:
```diff
@@ -0,0 +1,33 @@
+
+# This file is generated using `torch.nn.quantization.make_module`
```
put some comment with `@generated` - it makes Phabricator rendering nice, e.g.

```
# @generated from tools/autograd/templates/VariableType.cpp
```

dzhulgakov reviewed (changes requested):
Nicer!

raghuramank100 reviewed (changes requested):
Please address comments, thanks

raghuramank100 commented:
We need to have a test to ensure that the generated modules function correctly: For example, torch.add() and AddGen should be consistent and similarly for the quantized op and the modules

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

zafartahirov commented:
> Please add tests to ensure that the generated modules function correctly:
> 
> 1. torch.add() and AddGen should be consistent.
> 2. torch.ops.quantized.add() and QAddGen should be consistent.
> 3. QAddGen created via from_float() should match torch.ops.quantized.add() with the same arguments.

Will add tge tests in the next diff

jerryzh168 commented on torch/nn/quantized/modules/_generated.py:
```diff
@@ -0,0 +1,27 @@
+
+    def __init__(self, scale=1.0, zero_point=0):
+        super(Add, self).__init__()
+        self.scale = 1.0
```
I think these are supposed to be tensors?

raghuramank100 replied:
You need to assign them from init args

zafartahirov replied:
No, they cannot be tensors, because the ops expect double and long. However, if saving it as `buffer`, we might need to convert them to tensors, and use `.item()` when calling the op

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -229,6 +229,8 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
     QuantStub: nnq.Quantize,
     DeQuantStub: nnq.DeQuantize,
+    # Generated modules:
+    torch.quantization.Add: torch.quantization.QAdd,
```
why are these in quantization namespace?

raghuramank100 replied:
Yes, please update to nn.Add and nnq.Add

raghuramank100 commented on torch/quantization/tools/make_modules.py:
```diff
@@ -0,0 +1,110 @@
+
+    def __init__(self, scale=1.0, zero_point=0):
+        super({module_name}, self).__init__()
+        self.scale = 1.0
```
Please change to 
```
self.scale = scale
self.zero_point = zero_point
```

raghuramank100 reviewed (changes requested):
Please see comments

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

jerryzh168 commented on torch/nn/quantized/modules/_generated.py:
```diff
@@ -0,0 +1,30 @@
+            "Input float module must have observer attached"
+        assert (type(mod) == cls.__FLOAT_MODULE),\
+            "nnq.Add.from_float only works for " + cls.__FLOAT_MODULE.__name__
+        qparams = mod.observer.calculate_qparams()
```
nit: `act_scale, act_zp = mod.observer.calculate_qparams()` for better readability

zafartahirov replied:
there might be more parameters (i.e. dtype), which might make this a hassle to deal with

zafartahirov replied:
Resolved by adding `[:2]` in the `..._qparams()`

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -224,20 +223,22 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
     QuantStub: nnq.Quantize,
     DeQuantStub: nnq.DeQuantize,
+    # Generated modules:
+    nn.Add: nnq.quantized.Add,
```
nnq.Add?

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

raghuramank100 commented on torch/nn/quantized/modules/_generated.py:
```diff
@@ -0,0 +1,30 @@
+        assert (type(mod) == cls.__FLOAT_MODULE),\
+            "nnq.Add.from_float only works for " + cls.__FLOAT_MODULE.__name__
+        scale, zero_point = mod.observer.calculate_qparams()[:2]
+        mod = Add()
```
Should be:
 ```
mod = cls()
```

Commit:
Update on "make_module: First version"


Differential Revision: [D16455390](https://our.internmc.facebook.com/intern/diff/D16455390)

jerryzh168 reviewed (approved):
LGTM, we can change to the base class solution in future prs.
